tracing-log/src/lib.rs

```diff
@@ -91,6 +93,127 @@ pub trait AsTrace {
     fn as_trace(&self) -> Self::Trace;
 }
 
+/// Extends log `Event`s to provide complete `Metadata`.
+///
+/// In `tracing-log`, an `Event` produced by a log (through `AsTrace`) has an hard coded
+/// "log" target and no file, line, or module_path attributes. This happens because `Event`
+/// requires its `Metadata` to be `'static`, while log records provide them with a generic
+/// lifetime.
+///
+/// However, these values are stored in the `Event`'s fields and
+/// the [`normalized_metadata`] method allows to build a new `Metadata`
+/// that only lives as long as its source `Event`, but provides complete
+/// data.
+///
+/// It can typically be used by `Subscriber`s when processing an `Event`,
+/// to allow accessing its complete metadata in a consistent way,
+/// regardless of the source of its source.
+///
+/// [`normalized_metadata`]: trait.NormalizeEvent.html#normalized_metadata
+pub trait NormalizeEvent<'a> {
+    /// If this `Event` comes from a `log`, this method provides a new
+    /// normalized `Metadata` which has all available attributes
+    /// from the original log, including `file`, `line`, `module_path`
+    /// and `target`.
+    /// Returns `None` is the `Event` is not issued from a `log`.
+    fn normalized_metadata(&'a self) -> Metadata<'a>;
+    /// Returns wether this `Event` represents a log (from the `log` crate)
+    // FIXME: Not sure if should be part of the trait, could maybe
+    // be useful for subscribers?
+    fn is_log(&self) -> bool;
+}
+
+impl<'a> NormalizeEvent<'a> for Event<'a> {
+    fn normalized_metadata(&'a self) -> Metadata<'a> {
+        let original = self.metadata();
+        if self.is_log() {
+            let mut fields = LogVisitor::new_for(self);
+            self.record(&mut fields);
+
+            Metadata::new(
+                "log event",
+                fields.target.unwrap_or("log"),
+                original.level().clone(),
+                fields.file,
+                fields.line.map(|l| l as u32),
+                fields.module_path,
+                field::FieldSet::new(&["message"], original.callsite()),
+                Kind::EVENT,
+            )
+        } else {
+            Metadata::new(
+                original.name(),
+                original.target(),
+                original.level().clone(),
+                original.file(),
+                original.line(),
+                original.module_path(),
+                // FIXME how to get the original names array?
+                field::FieldSet::new(&[], original.callsite()),
```

Comment:
Do you know if there is a safe way to get the original `names` here?

Reply:
Ah, yeah, we can't just clone the original `FieldSet`, since it doesn't implement `Clone`. In that case...maybe we should just go back to returning an `Option` here.